### PR TITLE
csharp/circular-buffer: add mentor notes

### PR DESCRIPTION
**Pull request by fpetru**
I have added a new mentoring file on "Circular buffer" exercise. Please review it, and let me know your comments. Thank you !

### CODE REVIEW

**Comment by fpetru** [comment]
Thanks @NobbZ - yes, I will do a rebase.

File: tracks/csharp/exercises/circular-buffer/mentoring.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        _buffer.Enqueue(value);
+    }
+
+    public void Clear() => _buffer.Clear();
+}
+```
+
+### Common suggestions
+
+- While queue data structure could be implemented using a different type of collection (e.g. `Array<T>`), it is much easier to do this with `Queue<T>`.
```
Comment (ErikSchierboom) [inline tracks/csharp/exercises/circular-buffer/mentoring.md]:
```suggestion
- While the queue data structure could be implemented using a different type of collection (e.g. `Array<T>`), it is much easier to do this with `Queue<T>`. 
```

File: tracks/csharp/exercises/circular-buffer/mentoring.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    public void Clear() => _buffer.Clear();
+}
+```
+
+### Common suggestions
+
+- While queue data structure could be implemented using a different type of collection (e.g. `Array<T>`), it is much easier to do this with `Queue<T>`. 
+
+- To achieve a better performance, the queue should be initialized in the constructor (setting the initial capacity). This initialization would be preferred to small, and incremental memory allocations.
```
Comment (ErikSchierboom) [inline tracks/csharp/exercises/circular-buffer/mentoring.md]:
```suggestion
- To achieve better performance, the queue should be initialized in the constructor (setting the initial capacity). This initialization would be preferred to small, and incremental memory allocations.
```

Comment (ErikSchierboom) [reply tracks/csharp/exercises/circular-buffer/mentoring.md]: Great comment!

File: tracks/csharp/exercises/circular-buffer/mentoring.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+### Talking points
+
+- The class members (__capacity_ and __buffer_) should be made [readonly](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/readonly) - by doing this we could prevent changing their values outside of the constructor.
```
Comment (ErikSchierboom) [inline tracks/csharp/exercises/circular-buffer/mentoring.md]:
```suggestion
- The class fields should be made [readonly](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/readonly) - by doing this we could prevent changing their values outside of the constructor.
```

Comment (ErikSchierboom) [reply tracks/csharp/exercises/circular-buffer/mentoring.md]: I don't think we need to mention the fields.

File: tracks/csharp/exercises/circular-buffer/mentoring.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        if (_buffer.Count == _capacity) 
+            throw new InvalidOperationException("Circular buffer is full !");
+        else 
+            _buffer.Enqueue(value);   
+    }
+
+    public void Overwrite(T value)
+    {
+        if (_buffer.Count==_capacity) _buffer.Dequeue();
```
Comment (ErikSchierboom) [inline tracks/csharp/exercises/circular-buffer/mentoring.md]:
```suggestion
        if (_buffer.Count == _capacity) _buffer.Dequeue();
```

A little bit of consistency.

File: tracks/csharp/exercises/circular-buffer/mentoring.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        _capacity = capacity;
+        _buffer = new Queue<T>(capacity);
+    }
+
+    public T Read()
+    {
+        if (_capacity < 1) 
+            throw new InvalidOperationException("Circular buffer capacity needs to be greater than 0 !");
+        else
```
Comment (ErikSchierboom) [inline tracks/csharp/exercises/circular-buffer/mentoring.md]: You could consider losing the `else` statement.

File: tracks/csharp/exercises/circular-buffer/mentoring.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+            throw new InvalidOperationException("Circular buffer capacity needs to be greater than 0 !");
+        else 
+            return _buffer.Dequeue(); 
+    }
+
+    public void Write(T value)
+    {
+        if (_buffer.Count == _capacity) 
+            throw new InvalidOperationException("Circular buffer is full !");
+        else
```
Comment (ErikSchierboom) [inline tracks/csharp/exercises/circular-buffer/mentoring.md]: You could consider losing the `else` statement.

File: tracks/csharp/exercises/circular-buffer/mentoring.md (hunk 5 above)
Comment (fpetru) [reply tracks/csharp/exercises/circular-buffer/mentoring.md]: yes, removed.

File: tracks/csharp/exercises/circular-buffer/mentoring.md (hunk 6 above)
Comment (fpetru) [reply tracks/csharp/exercises/circular-buffer/mentoring.md]: yes, removed.

**Comment by fpetru** [comment]
Thank you @ErikSchierboom for your review. I've applied all changes, and also further extended with an additional solution & notes.

**Comment by fpetru** [comment]
@ErikSchierboom @NobbZ - whenever this PR is accepted, could you help me to "squash and merge" the earlier commits ? I've tried to rebase it, but it did not work so well. I will make sure the next PR to be more clean.

**Comment by NobbZ** [comment]
As soon as @ErikSchierboom has resolved all discussions/accepted the changes made, the person who merges should just use the "squash and merge" option on the Web-UI.

**Pull request merged**